Documentation/design-docs/tailcalls-with-helpers.md

```diff
@@ -0,0 +1,375 @@
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+```
+### In CallTarget thunk for D, about to call D
+The thunk will now extract parameters for D from the thread local storage and call D.
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+```
+### In D
+We are in the last function of the chain, so after it does its work, it returns to its CallTarget thunk. 
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+* Stack arguments of D
+* Return address of D
+* Callee saved registers and locals of D
+```
+### Returned to CallTarget thunk for D with ChainCall flag in the TailCallHelperStackEntry **clear**
+The thunk checks the ChainCall flag and since it is clear, it recognizes we are now returning from the call chain and so it returns the result of the D. 
+```
+* Return address of A
+* Callee saved registers and locals of A
+* Stack arguments of B
+* Return address of B
+* Callee saved registers and locals of CallTarget thunk for D
+```
+### Returned to A
+We are back in A and we have the return value of the call chain.
+```
+* Return address of A
+* Callee saved registers and locals of A
+```
+
+## Example of thunks generated for a simple generic method
+
+```C#
+struct Point
+{
+    public int x;
+    public int y;
+    public int z;
+}
+
+class Foo
+{
+    public Point Test<T>(int x, T t) where T : class
+    {
+        Console.WriteLine("T: {0}", typeof(t));
+        return new Point(x, x, x);
+    }
+}
+```
+
+For tail calling the Test function, the IL helpers could look e.g. as follows:
+
+```IL
+.method public static Point StoreArgumentsTest(object thisObj, object genericContext, int x, object t) cil managed
+{
+	.maxstack 4
+	.locals init (
+		[0] valuetype Point,
+	)
+
+    call native int FetchArgumentBuffer()
+
+	dup
+	ldarg.0 // "thisObj"
+	stobj object
+	sizeof object
+	add
+	dup
+
+    ldarg.1 // "genericContext"
+	stobj object
+	sizeof object
+	add
+	dup
+
+	ldarg.2 // "x"
+	stind.i
+	sizeof int32
+	add
+	dup
+
+    ldarg.3 // "t"
+	stobj object
+
+	ldloca.s 0 // return default(Point)
+	initobj Point
+	ldloc.0
+
+	ret
+}
+
+```
+
+```IL
+.method public static Point CallTargetTest() cil managed
+{
+    .maxstack 4
+    .locals init (
+       [0] native int buffer
+    )
+    call native int FetchArgumentBuffer()
+    stloc.0
+
+    ldloc.0
+    ldobj object // this
+    ldloc.0
+    sizeof object
+    add
+    stloc.0
+
+    ldloc.0
+    ldobj object // generic context
+    ldloc.0
+    sizeof object
+    add
+    stloc.0
+
+    ldloc.0
+    ldind.i // int x
+    ldloc.0
+    sizeof int32
+    add
+    stloc.0
+
+    ldloc.0
+    ldobj object // T t
+
+    calli Point (object, object, int32, object)
```

Comment:
Should this be regular call instead? (I do not see where the address for calli is coming from.)

Reply:
I know this is years old at this point, but it seems hard (/potentially fragile) to make this `calli`. For example, on x86 the type context comes after user args, so this would need to be special cased there in the JIT and the AOT scenario.

I am not particularly sure I understand why this cannot just be a direct `call` (which then may also be potentially inlined). There is a large discussion about it above, but I am not sure I understand why this should not be technically possible. After all the `CallTargetThunk` can just "semantically" represent the tail of the method except it gets its arguments from somewhere else.

Reply:
@jakobbotsch I'm not sure how we could make it work without the calli. When making a tail call with helper, we first call the StoreArguments which stores all the arguments including the "this" pointer and the generic pointer argument. The CallTarget then gets the buffer with the stored arguments, extracts them and passes them to the target function. If we used call instead of calli here, how could we pass in the extracted generic argument?

Reply:
I can see why we may have problems since we somehow will have to pass the generic context through to the call thunk, and we may not have enough stack space to do this (only a problem on x86). I wonder if this could be special cased in the x86 helper instead, by considering it to have __fastcall calling convention. However it would have to push it on the stack before calling the call thunk.

Reply:
On other platforms/ABIs this is what I would imagine, taking @jkotas's example from above:
```csharp
static void foo<T>()
{
    ...
    tailcall.    bar<List<T>>(1,2,3,4,5,6,7,8);
}

static void bar<T>(int a1, int a2, int a3, int a4, int a5, int a6, int a7, int a8)
{
   ....
}
```

We turn this into:
```
static void foo<T>()
{
    ...
    call StoreBarArguments(1, 2, 3, 4, 5, 6, 7, 8);
    tailcall.    CallBar<T>();
}

static void CallBar<T>()
{
    bar<List<T>>(args...);
}
```

and JIT treats the tailcall to `CallBar` exactly as normal except it uses the tail call helper as its target. On other ABIs the tailcall to CallBar here is always possible.